### PR TITLE
Avanti auth override support

### LINKED ISSUE #364: Support authentication for plio using third party auth tokens

### PR DESCRIPTION
**Pull request by deepansh96**
Fixes #364
Depends on [this PR](https://github.com/avantifellows/plio-backend/pull/207)

## Test Plan

- [ ] ~Test Responsiveness~
   - [ ] Laptop (1200px)
   - [ ] Tablet (760px)
   - [ ] Phone (320px)
- [ ] ~Cross-Browser Testing~
   - [ ] Chrome
   - [ ] Firefox
   - [ ] Safari
- [ ] ~Local Language Support~
- [ ] Wrote tests
- [x] Tested locally
- [ ] Tested on staging
- [ ] Tested on an actual physical phone
- [ ] Tested on production
- [x] Check for bundle size [here](https://bundlephobia.com/) if adding a package
- [x] Added relevant details like Labels/Projects/Milestones etc.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -29,7 +29,7 @@
     "vue-gtag": "^2.0.0",
     "vue-i18n": "^9.0.0",
     "vue-inline-svg": "^3.0.0-beta.2",
-    "vue-router": "^4.0.0-beta.12",
+    "vue-router": "^4.0.10",
```
Comment (dalmia) [inline package.json:32]: any specific reason for this?

File: src/App.vue [hunk 2]
```diff
@@ -173,9 +173,19 @@ export default {
...
-        if (!isUserInWorkspace) this.$router.replace({ name: "Home" });
+
+        if (!isUserInWorkspace) {
+          // make sure to pass the query params as it is so they're not lost
+          // while redirecting
+          this.$router.replace({
+            name: "Home",
+            query: this.$route.query,
+          });
+        }
```
Comment (dalmia) [inline src/App.vue:188]: please clarify why this change was needed.

File: src/router/index.js [hunk 3]
```diff
@@ -9,6 +9,8 @@ import i18n from "@/services/Localisation/i18n.js";
 import { useToast } from "vue-toastification";
 
 const toast = useToast();
+// these keys should be present in the third party auth query request URL
```
Comment (dalmia) [inline src/router/index.js:12]:
```suggestion
// these keys should be present as query params when a third party
// auth is used to access a router 
```

File: src/router/index.js [hunk 4]
```diff
@@ -100,7 +105,7 @@ The code below works on `isAuthenticated` state and before every route:
 2. Redirects user to home if user is already logged in and visiting a page that is intended for guest (route.meta.guest)
 */
 
-router.beforeEach((to, from, next) => {
+router.beforeEach((to, from) => {
```
Comment (dalmia) [inline src/router/index.js:108]: why remove this?

File: src/router/index.js [hunk 5]
```diff
@@ -115,55 +120,55 @@ router.beforeEach((to, from, next) => {
...
-function restrictUnapprovedUser(to, from, next) {
-  if (store.getters["auth/isUserApproved"]) {
-    next();
-    return;
-  }
-  next({ name: "Home", params: { org: to.params.org }, replace: true });
+function restrictUnapprovedUser(to) {
+  if (!store.getters["auth/isUserApproved"])
+    return { name: "Home", params: { org: to.params.org }, replace: true };
+  else return;
```
Comment (dalmia) [inline src/router/index.js:171]: didn't understand the significance of this change.

File: src/router/index.js [hunk 6]
```diff
@@ -115,55 +120,55 @@ router.beforeEach((to, from, next) => {
...
 // set organization in vuex state if the route org parameter is in vuex user organizations array
-router.beforeEach((to, from, next) => {
-  if (!store.getters["auth/isAuthenticated"]) {
-    next();
-    return;
+router.beforeEach((to) => {
+  if (store.getters["auth/isAuthenticated"]) {
+    if (to.params.org != "" && to.params.org != undefined)
+      store.dispatch("auth/setActiveWorkspace", to.params.org);
+    else store.dispatch("auth/unsetActiveWorkspace");
```
Comment (dalmia) [inline src/router/index.js:163]: same question as before - maybe there's just one common theme who changes have been applied everywhere.

File: src/router/index.js [hunk 7]
```diff
@@ -115,55 +120,55 @@ router.beforeEach((to, from, next) => {
...
-    if (store.getters["auth/isAuthenticated"]) {
-      next({ name: "Home" });
-      return;
-    }
-    next();
-  } else {
-    next();
-  }
+    if (store.getters["auth/isAuthenticated"]) return { name: "Home" };
+  } else return;
```
Comment (dalmia) [inline src/router/index.js:151]: same question

File: src/router/index.js [hunk 8]
```diff
@@ -115,55 +120,55 @@ router.beforeEach((to, from, next) => {
   }
 
   if (to.matched.some((record) => record.meta.requiresAuth)) {
-    if (store.getters["auth/isAuthenticated"]) {
-      next();
+    // app has to be authenticated using third party auth if all the query params
+    // match the keys in `requiredAuthKeys` and they're not empty or undefined
+    var isThirdPartyAuth =
+      requiredAuthKeys.every((key) => Object.keys(to.query).includes(key)) &&
```
Comment (dalmia) [inline src/router/index.js:126]: Object.keys(to.query) -> can be saved in a variable? - being estimated every time for each key

File: src/pages/Player.vue [hunk 9]
```diff
@@ -232,6 +234,33 @@ export default {
     },
   },
   async created() {
+    // creating a promise for the third party auth functionality
+
```
Comment (dalmia) [inline src/pages/Player.vue:238]: best to remove this - a bit confusing - makes the top line seem like a description for the created function.

File: src/pages/Player.vue [hunk 10]
```diff
@@ -232,6 +234,33 @@ export default {
...
+    // if the app needs to authenticate via third party, resolve this promise only when all tasks are done
+    // and the authenticated user is set
+
```
Comment (dalmia) [inline src/pages/Player.vue:241]: same as above.

File: src/router/index.js (hunk 4 above)
Comment (deepansh96) [reply src/router/index.js:108]: https://next.router.vuejs.org/guide/advanced/navigation-guards.html#optional-third-argument-next

File: src/router/index.js [hunk 11]
```diff
@@ -123,9 +123,10 @@ router.beforeEach((to, from) => {
   if (to.matched.some((record) => record.meta.requiresAuth)) {
     // app has to be authenticated using third party auth if all the query params
     // match the keys in `requiredAuthKeys` and they're not empty or undefined
-    var isThirdPartyAuth =
-      requiredAuthKeys.every((key) => Object.keys(to.query).includes(key)) &&
-      Object.keys(to.query).every(
+    let query_params = Object.keys(to.query);
```
Comment (dalmia) [inline src/router/index.js:126]: queryParams

File: src/pages/Player.vue [hunk 12]
```diff
@@ -245,16 +245,34 @@ export default {
     });
 
     if (this.isThirdPartyAuth) {
+      console.log("if enter");
```
Comment (dalmia) [inline src/pages/Player.vue:248]:
```suggestion
```

File: src/pages/Player.vue [hunk 13]
```diff
@@ -245,16 +245,34 @@ export default {
...
         unique_id: this.thirdPartyUniqueId,
         access_token: this.thirdPartyAccessToken,
-      }).then(async (response) => {
-        await this.setAccessToken(response.data);
-        thirdPartyAuthPromiseResolve();
-      });
+      })
+        .then(async (response) => {
+          await this.setAccessToken(response.data);
+          console.log("then");
```
Comment (dalmia) [inline src/pages/Player.vue:258]:
```suggestion
```
